Mark `simplify_aggregate_to_copy` mir-opt as unsound

jieyouxu opened the pull request:
Mark the `simplify_aggregate_to_copy` mir-opt added in #128299 as unsound as it seems to miscompile the MCVE reported in https://github.com/rust-lang/rust/issues/132353. The mir-opt can be re-enabled once this case is fixed.

```rs
fn pop_min(mut score2head: Vec<Option<usize>>) -> Option<usize> {
    loop {
        if let Some(col) = score2head[0] {
            score2head[0] = None;
            return Some(col);
        }
    }
}

fn main() {
    let min = pop_min(vec![Some(1)]);
    println!("min: {:?}", min);
    // panic happens here on beta in release mode
    // but not in debug mode
    min.unwrap();
}
```

This MCVE is included as a `run-pass` ui regression test in the first commit. I built the ui test with a nightly manually, and can reproduce the behavioral difference with `-C opt-level=0` and `-C opt-level=1`. Locally, this ui test will fail unless it was run on a compiler built with the second commit marking the mir-opt as unsound thus disabling it by default.

This PR **partially reverts** commit e7386b3, reversing changes made to 02b1be1. The mir-opt implementation is just marked as unsound but **not** reverted to make reland reviews easier. Test changes are **reverted if they were not pure additions**. Tests added by the original PR received `-Z unsound-mir-opts` compile-flags.

cc @DianQK @cjgillot (PR author and reviewer of #128299)

jieyouxu committed:
Add a regression test for #132353

To catch at least one pattern that was miscompiled. The test is a
minimization of the MCVE reported in
<https://github.com/rust-lang/rust/issues/132353>.

jieyouxu commented:
Nominating for beta-backport as this is a mir-opt miscompile.
@rustbot label +beta-nominated

DianQK commented on tests/codegen/clone_as_copy.rs:
```diff
@@ -1,40 +0,0 @@
-//@ revisions: DEBUGINFO NODEBUGINFO
-//@ compile-flags: -O -Cno-prepopulate-passes
```
I think you can also add `-Zunsound-mir-opts` here.

jieyouxu replied:
@DianQK is there any particular "markers" that will make this easier to find? Or is `-Zunsound-mir-opts` alone sufficient?

DianQK replied:
> @DianQK is there any particular "markers" that will make this easier to find? Or is `-Zunsound-mir-opts` alone sufficient?

Just add it, I think: `-O -Cno-prepopulate-passes -Zunsound-mir-opts`.

jieyouxu replied:
Added `-Zunsound-mir-opts` along with FIXMEs linking to original issue.

DianQK commented on tests/mir-opt/gvn_clone.rs:
```diff
@@ -1,17 +0,0 @@
-//@ test-mir-pass: GVN
-//@ compile-flags: -Zmir-enable-passes=+InstSimplify-before-inline
```
Here as well.

jieyouxu replied:
Added `-O -Cno-prepopulate-passes -Zunsound-mir-opts` + FIXME.

DianQK commented on tests/codegen/enum/unreachable_enum_default_branch.rs:
```diff
@@ -28,13 +28,11 @@ pub fn implicit_match(x: Int) -> bool {
 #[no_mangle]
 pub fn if_let(val: Result<i32, ()>) -> Result<i32, ()> {
-    if let Ok(x) = val { Ok(x * 2) } else { Err(()) }
+    if let Ok(x) = val { Ok(x) } else { Err(()) }
```
Did you just use revert? This change can retain.

jieyouxu replied:
Yes, that's what I did for the tests. I can address the tests as suitable.

jieyouxu replied:
Kept the changes to this file.

jieyouxu committed:
Mark `simplify_aggregate_to_copy` mir-opt as unsound

Co-authored-by: DianQK <dianqk@dianqk.net>

jieyouxu commented:
(Accidentally added too many flags, oops)

jieyouxu commented on tests/mir-opt/pre-codegen/try_identity.rs:
```diff
@@ -1,3 +1,6 @@
+//@ compile-flags: -Zunsound-mir-opts
+// FIXME: see <https://github.com/rust-lang/rust/issues/132353>
+
```
Problem [NOT-PURE-ADD 2/N]: and this as well

jieyouxu replied:
Dropped `-Zunsound-mir-opts`

jieyouxu commented on tests/mir-opt/pre-codegen/no_inlined_clone.rs:
```diff
@@ -1,3 +1,6 @@
+//@ compile-flags: -Zunsound-mir-opts
+// FIXME: see <https://github.com/rust-lang/rust/issues/132353>
```
Dropped `-Zunsound-mir-opts`

jieyouxu commented on tests/mir-opt/pre-codegen/vec_deref.rs:
```diff
@@ -1,3 +1,5 @@
+//@ compile-flags: -Zunsound-mir-opts
+// FIXME: see <https://github.com/rust-lang/rust/issues/132353>
```
Dropped `-Zunsound-mir-opts`

jieyouxu commented:
Changes since last review:

- Added `-Zunsound-mir-opts` *only* to tests that are pure additions in https://github.com/rust-lang/rust/pull/128299.
- Retained changes to `tests/codegen/enum/unreachable_enum_default_branch.rs`.

jieyouxu commented:
Probably r? mir-opt

cjgillot reviewed (commented):
Indeed, we aren't allowed to introduce new dereferences without checking for aliasing.
r=me

jieyouxu commented on tests/ui/mir/clone-canonicalization-miscompile-132353.rs:
```diff
@@ -0,0 +1,25 @@
+// reproduce the miscompile.
+//@[release] compile-flags: -C opt-level=1
+//@[debug] compile-flags: -C opt-level=0
+//@ run-pass
```
To clarify, do you mean retain this ui test (for now), but also create a new mir-opt test which shows the GVN diff?

jieyouxu replied:
Thanks, I tried to add a `GVN.diff` test w/ this snippet, lmw if the test needs adjustments.

jieyouxu committed:
Add a mir-opt GVN test for #128299

Co-authored-by: DianQK <dianqk@dianqk.net>

jieyouxu commented:
Added a GVN diff mir-opt test as requested.
@rustbot ready

bors commented:
:pushpin: Commit 29db55eab7f6c7aeb2f82dbb07a1d4623b23d1dd has been approved by `cjgillot`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cjgillot 29db55eab7f6c7aeb2f82dbb07a1d4623b23d1dd -->
<!-- homu: {"type":"Approved","sha":"29db55eab7f6c7aeb2f82dbb07a1d4623b23d1dd","approver":"cjgillot","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

cjgillot commented:
Thanks!

wesleywiser commented:
@bors p=1

Mitigates unsound MIR opt

bors commented:
:hourglass: Testing commit 29db55eab7f6c7aeb2f82dbb07a1d4623b23d1dd with merge 2b66b9b5a3e592b01a5f8c2606aa2b566e601494...
<!-- homu: {"type":"BuildStarted","head_sha":"29db55eab7f6c7aeb2f82dbb07a1d4623b23d1dd","merge_sha":"2b66b9b5a3e592b01a5f8c2606aa2b566e601494"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-19`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/11600761901/job/32301993191) [(plain)](https://github.com/rust-lang-ci/rust/commit/2b66b9b5a3e592b01a5f8c2606aa2b566e601494/checks/32301993191/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
failures:

---- [codegen] tests/codegen/try_question_mark_nop.rs#NINETEEN stdout ----

error in revision `NINETEEN`: verification with 'FileCheck' failed
status: exit status: 1
command: "/usr/lib/llvm-19/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/try_question_mark_nop.NINETEEN/try_question_mark_nop.ll" "/checkout/tests/codegen/try_question_mark_nop.rs" "--check-prefix=CHECK" "--check-prefix" "NINETEEN" "--check-prefix" "NONMSVC" "--allow-unused-prefixes" "--dump-input-context" "100"
--- stderr -------------------------------
--- stderr -------------------------------
/checkout/tests/codegen/try_question_mark_nop.rs:19:20: error: NINETEEN-NEXT: expected string not found in input
 // NINETEEN-NEXT: [[TRUNC:%.*]] = trunc nuw i32 %0 to i1
                   ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/try_question_mark_nop.NINETEEN/try_question_mark_nop.ll:8:7: note: scanning from here
      ^
      ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/try_question_mark_nop.NINETEEN/try_question_mark_nop.ll:9:24: note: possible intended match here
 %2 = insertvalue { i32, i32 } poison, i32 %0, 0


Input file: /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/try_question_mark_nop.NINETEEN/try_question_mark_nop.ll


-dump-input=help explains the following input dump.
Input was:
<<<<<<
<<<<<<
           1: ; ModuleID = 'try_question_mark_nop.e67e5ea31bdd6ce8-cgu.0' 
           2: source_filename = "try_question_mark_nop.e67e5ea31bdd6ce8-cgu.0" 
           3: target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128" 
           4: target triple = "x86_64-unknown-linux-gnu" 
           5:  
           6: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(none) uwtable 
           7: define { i32, i32 } @option_nop_match_32(i32 noundef range(i32 0, 2) %0, i32 %1) unnamed_addr #0 { 
           8: start: 
next:19'0           X error: no match found
           9:  %2 = insertvalue { i32, i32 } poison, i32 %0, 0 
next:19'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
next:19'1                            ?                          possible intended match
          10:  %3 = insertvalue { i32, i32 } %2, i32 %1, 1 
next:19'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          11:  ret { i32, i32 } %3 
next:19'0     ~~~~~~~~~~~~~~~~~~~~~
          12: } 
next:19'0     ~~
next:19'0     ~
next:19'0     ~
          14: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(none) uwtable 
next:19'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          15: define { i32, i32 } @option_nop_traits_32(i32 noundef range(i32 0, 2) %0, i32 %1) unnamed_addr #0 { 
next:19'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          16: start: 
          17:  %2 = insertvalue { i32, i32 } poison, i32 %0, 0 
          18:  %3 = insertvalue { i32, i32 } %2, i32 %1, 1 
          19:  ret { i32, i32 } %3 
          20: } 
          21:  
          22: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(none) uwtable 
          23: define { i32, i32 } @result_nop_match_32(i32 noundef range(i32 0, 2) %0, i32 noundef %1) unnamed_addr #0 { 
          24: start: 
          25:  %2 = insertvalue { i32, i32 } poison, i32 %0, 0 
          26:  %3 = insertvalue { i32, i32 } %2, i32 %1, 1 
          27:  ret { i32, i32 } %3 
          28: } 
          29:  
          30: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(none) uwtable 
          31: define { i32, i32 } @result_nop_traits_32(i32 noundef range(i32 0, 2) %0, i32 noundef %1) unnamed_addr #0 { 
          32: start: 
          33:  %2 = insertvalue { i32, i32 } poison, i32 %0, 0 
          34:  %3 = insertvalue { i32, i32 } %2, i32 %1, 1 
          35:  ret { i32, i32 } %3 
          36: } 
          37:  
          38: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(none) uwtable 
          39: define { i64, i64 } @result_nop_match_64(i64 noundef range(i64 0, 2) %0, i64 noundef %1) unnamed_addr #0 { 
          40: start: 
          41:  %2 = insertvalue { i64, i64 } poison, i64 %0, 0 
          42:  %3 = insertvalue { i64, i64 } %2, i64 %1, 1 
          43:  ret { i64, i64 } %3 
          44: } 
          45:  
          46: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(none) uwtable 
          47: define { i64, i64 } @result_nop_traits_64(i64 noundef range(i64 0, 2) %0, i64 noundef %1) unnamed_addr #0 { 
          48: start: 
          49:  %2 = insertvalue { i64, i64 } poison, i64 %0, 0 
          50:  %3 = insertvalue { i64, i64 } %2, i64 %1, 1 
          51:  ret { i64, i64 } %3 
          52: } 
          53:  
          54: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(none) uwtable 
          55: define { i64, ptr } @result_nop_match_ptr(i64 noundef range(i64 0, 2) %0, ptr noundef %1) unnamed_addr #0 { 
          56: start: 
          57:  %2 = insertvalue { i64, ptr } poison, i64 %0, 0 
          58:  %3 = insertvalue { i64, ptr } %2, ptr %1, 1 
          59:  ret { i64, ptr } %3 
          60: } 
          61:  
          62: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(none) uwtable 
          63: define { i64, ptr } @result_nop_traits_ptr(i64 noundef range(i64 0, 2) %0, ptr noundef %1) unnamed_addr #0 { 
          64: start: 
          65:  %2 = insertvalue { i64, ptr } poison, i64 %0, 0 
          66:  %3 = insertvalue { i64, ptr } %2, ptr %1, 1 
          67:  ret { i64, ptr } %3 
          68: } 
          69:  
          70: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(none) uwtable 
          71: define { i32, i32 } @control_flow_nop_match_32(i32 noundef range(i32 0, 2) %0, i32 noundef %1) unnamed_addr #0 { 
          72: start: 
          73:  %2 = insertvalue { i32, i32 } poison, i32 %0, 0 
          74:  %3 = insertvalue { i32, i32 } %2, i32 %1, 1 
          75:  ret { i32, i32 } %3 
          76: } 
          77:  
          78: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(none) uwtable 
          79: define { i32, i32 } @control_flow_nop_traits_32(i32 noundef range(i32 0, 2) %0, i32 noundef %1) unnamed_addr #0 { 
          80: start: 
          81:  %2 = insertvalue { i32, i32 } poison, i32 %0, 0 
          82:  %3 = insertvalue { i32, i32 } %2, i32 %1, 1 
          83:  ret { i32, i32 } %3 
          84: } 
          85:  
          86: attributes #0 = { mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(none) uwtable "probe-stack"="inline-asm" "target-cpu"="x86-64" } 
          87:  
          88: !llvm.module.flags = !{!0, !1} 
          89: !llvm.ident = !{!2} 
          90:  
          91: !0 = !{i32 8, !"PIC Level", i32 2} 
          92: !1 = !{i32 2, !"RtLibUseGOT", i32 1} 
          93: !2 = !{!"rustc version 1.84.0-nightly (2b66b9b5a 2024-10-30)"} 
------------------------------------------



```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/11600761901/job/32303302804)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/11600761901/job/32303302804","builder_name":"checks-actions"} -->

DianQK commented:
For me, both of reverting `try_question_mark_nop.rs` or adding `-Zunsound-mir-opts` are fine. :)

jieyouxu commented:
> For me, both of reverting `try_question_mark_nop.rs` or adding `-Zunsound-mir-opts` are fine. :)

Reverted changes to `try_question_mark_nop.rs` for now because it wasn't a pure addition.

DianQK commented on tests/codegen/try_question_mark_nop.rs:
```diff
@@ -1,25 +1,22 @@
 use std::ops::ControlFlow::{self, Break, Continue};
 use std::ptr::NonNull;
 
+// FIXME: The `trunc` and `select` instructions can be eliminated.
```
This introduced by `simplify_aggregate_to_copy`, so you can just remove it.

jieyouxu replied:
Removed.

DianQK commented on tests/codegen/try_question_mark_nop.rs:
```diff
@@ -1,25 +1,22 @@
-    // NINETEEN-NEXT: [[FIRST:%.*]] = select i1 [[TRUNC]], i32 %0
-    // NINETEEN-NEXT: insertvalue { i32, i32 } poison, i32 [[FIRST]], 0
-    // TWENTY-NEXT: insertvalue { i32, i32 } poison, i32 %0, 0
+    // CHECK-NEXT: [[TRUNC:%.*]] = trunc nuw i32 %0 to i1
+    // CHECK-NEXT: [[FIRST:%.*]] = select i1 [[TRUNC]], i32 %0
+    // CHECK-NEXT: insertvalue { i32, i32 } poison, i32 [[FIRST]]
```
```suggestion
    // CHECK-NEXT: insertvalue { i32, i32 } poison, i32 %0, 0
```

jieyouxu replied:
Changed the matchers to

```rs
// CHECK-NEXT: [[REG1:%.*]] = insertvalue { i32, i32 } poison, i32 %0, 0
// CHECK-NEXT: [[REG2:%.*]] = insertvalue { i32, i32 } [[REG1]], i32 %1, 1
// CHECK-NEXT: ret { i32, i32 } [[REG2]]
```

to make the test pass.

dianqk reviewed (commented):
I think you can r=cjgillot after the new suggesting changes. :3

jieyouxu commented:
Only change was the `try_question_mark_nop.rs` test, passes locally, so

@bors r=cjgillot,DianQK rollup=iffy

bors commented:
:pushpin: Commit cfb4c05d77df4a6bcc53924eddb3a35102b65da9 has been approved by `cjgillot,DianQK`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cjgillot,DianQK cfb4c05d77df4a6bcc53924eddb3a35102b65da9 -->
<!-- homu: {"type":"Approved","sha":"cfb4c05d77df4a6bcc53924eddb3a35102b65da9","approver":"cjgillot,DianQK","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit cfb4c05d77df4a6bcc53924eddb3a35102b65da9 with merge 99b0f2109fa624a17049ee07aafe7b6a214c4cfe...
<!-- homu: {"type":"BuildStarted","head_sha":"cfb4c05d77df4a6bcc53924eddb3a35102b65da9","merge_sha":"99b0f2109fa624a17049ee07aafe7b6a214c4cfe"} -->

rust-log-analyzer commented:
The job **`dist-x86_64-msvc`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/11609160216/job/32325892140) [(plain)](https://github.com/rust-lang-ci/rust/commit/99b0f2109fa624a17049ee07aafe7b6a214c4cfe/checks/32325892140/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[2024-10-31T10:05:09Z DEBUG collector::compile::benchmark] Benchmark iteration 1/1
[2024-10-31T10:05:09Z INFO  collector::compile::execute] run_rustc with incremental=false, profile=Check, scenario=Some(Full), patch=None, backend=Llvm, phase=benchmark
[2024-10-31T10:05:09Z DEBUG collector::compile::execute] "\\\\?\\C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage0\\bin\\cargo.exe" "rustc" "--manifest-path" "Cargo.toml" "-p" "path+file:///C:/a/_temp/msys64/tmp/.tmpIADZiT#ctfe-stress-5@0.1.0" "--profile" "check" "--" "--wrap-rustc-with" "Eprintln"
[2024-10-31T10:05:15Z INFO  collector::compile::execute] run_rustc with incremental=true, profile=Check, scenario=Some(IncrFull), patch=None, backend=Llvm, phase=benchmark
[2024-10-31T10:05:15Z DEBUG collector::compile::execute] "\\\\?\\C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage0\\bin\\cargo.exe" "rustc" "--manifest-path" "Cargo.toml" "-p" "path+file:///C:/a/_temp/msys64/tmp/.tmpIADZiT#ctfe-stress-5@0.1.0" "--profile" "check" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=C:\\a\\_temp\\msys64\\tmp\\.tmpIADZiT\\incremental-state"
[2024-10-31T10:05:21Z INFO  collector::compile::execute] run_rustc with incremental=true, profile=Check, scenario=Some(IncrUnchanged), patch=None, backend=Llvm, phase=benchmark
[2024-10-31T10:05:21Z DEBUG collector::compile::execute] "\\\\?\\C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage0\\bin\\cargo.exe" "rustc" "--manifest-path" "Cargo.toml" "-p" "path+file:///C:/a/_temp/msys64/tmp/.tmpIADZiT#ctfe-stress-5@0.1.0" "--profile" "check" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=C:\\a\\_temp\\msys64\\tmp\\.tmpIADZiT\\incremental-state"
[2024-10-31T10:05:21Z DEBUG collector::compile::benchmark] Benchmark iteration 1/1
[2024-10-31T10:05:21Z INFO  collector::compile::execute] run_rustc with incremental=false, profile=Debug, scenario=Some(Full), patch=None, backend=Llvm, phase=benchmark
[2024-10-31T10:05:21Z DEBUG collector::compile::execute] "\\\\?\\C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage0\\bin\\cargo.exe" "rustc" "--manifest-path" "Cargo.toml" "-p" "path+file:///C:/a/_temp/msys64/tmp/.tmpoYvyBw#ctfe-stress-5@0.1.0" "--" "--wrap-rustc-with" "Eprintln"
[2024-10-31T10:05:27Z INFO  collector::compile::execute] run_rustc with incremental=true, profile=Debug, scenario=Some(IncrFull), patch=None, backend=Llvm, phase=benchmark
[2024-10-31T10:05:27Z INFO  collector::compile::execute] run_rustc with incremental=true, profile=Debug, scenario=Some(IncrFull), patch=None, backend=Llvm, phase=benchmark
[2024-10-31T10:05:27Z DEBUG collector::compile::execute] "\\\\?\\C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage0\\bin\\cargo.exe" "rustc" "--manifest-path" "Cargo.toml" "-p" "path+file:///C:/a/_temp/msys64/tmp/.tmpoYvyBw#ctfe-stress-5@0.1.0" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=C:\\a\\_temp\\msys64\\tmp\\.tmpoYvyBw\\incremental-state"
[2024-10-31T10:05:33Z INFO  collector::compile::execute] run_rustc with incremental=true, profile=Debug, scenario=Some(IncrUnchanged), patch=None, backend=Llvm, phase=benchmark
[2024-10-31T10:05:33Z DEBUG collector::compile::execute] "\\\\?\\C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage0\\bin\\cargo.exe" "rustc" "--manifest-path" "Cargo.toml" "-p" "path+file:///C:/a/_temp/msys64/tmp/.tmpoYvyBw#ctfe-stress-5@0.1.0" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=C:\\a\\_temp\\msys64\\tmp\\.tmpoYvyBw\\incremental-state"
[2024-10-31T10:05:33Z DEBUG collector::compile::benchmark] Benchmark iteration 1/1
[2024-10-31T10:05:34Z INFO  collector::compile::execute] run_rustc with incremental=false, profile=Opt, scenario=Some(Full), patch=None, backend=Llvm, phase=benchmark
[2024-10-31T10:05:34Z DEBUG collector::compile::execute] "\\\\?\\C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage0\\bin\\cargo.exe" "rustc" "--manifest-path" "Cargo.toml" "-p" "path+file:///C:/a/_temp/msys64/tmp/.tmpfqlnbp#ctfe-stress-5@0.1.0" "--release" "--" "--wrap-rustc-with" "Eprintln"
[2024-10-31T10:05:39Z INFO  collector::compile::execute] run_rustc with incremental=true, profile=Opt, scenario=Some(IncrFull), patch=None, backend=Llvm, phase=benchmark
[2024-10-31T10:05:39Z INFO  collector::compile::execute] run_rustc with incremental=true, profile=Opt, scenario=Some(IncrFull), patch=None, backend=Llvm, phase=benchmark
[2024-10-31T10:05:39Z DEBUG collector::compile::execute] "\\\\?\\C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage0\\bin\\cargo.exe" "rustc" "--manifest-path" "Cargo.toml" "-p" "path+file:///C:/a/_temp/msys64/tmp/.tmpfqlnbp#ctfe-stress-5@0.1.0" "--release" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=C:\\a\\_temp\\msys64\\tmp\\.tmpfqlnbp\\incremental-state"
[2024-10-31T10:05:45Z INFO  collector::compile::execute] run_rustc with incremental=true, profile=Opt, scenario=Some(IncrUnchanged), patch=None, backend=Llvm, phase=benchmark
[2024-10-31T10:05:45Z DEBUG collector::compile::execute] "\\\\?\\C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage0\\bin\\cargo.exe" "rustc" "--manifest-path" "Cargo.toml" "-p" "path+file:///C:/a/_temp/msys64/tmp/.tmpfqlnbp#ctfe-stress-5@0.1.0" "--release" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=C:\\a\\_temp\\msys64\\tmp\\.tmpfqlnbp\\incremental-state"
Executing benchmark diesel-1.4.8 (4/8)
Preparing diesel-1.4.8
[2024-10-31T10:05:46Z INFO  collector::compile::execute] run_rustc with incremental=false, profile=Debug, scenario=None, patch=None, backend=Llvm, phase=dependencies
[2024-10-31T10:05:46Z INFO  collector::compile::execute] run_rustc with incremental=false, profile=Check, scenario=None, patch=None, backend=Llvm, phase=dependencies
---
[2024-10-31T10:07:05Z DEBUG collector::compile::benchmark] Benchmark iteration 1/1
[2024-10-31T10:07:05Z INFO  collector::compile::execute] run_rustc with incremental=false, profile=Check, scenario=Some(Full), patch=None, backend=Llvm, phase=benchmark
[2024-10-31T10:07:05Z DEBUG collector::compile::execute] "\\\\?\\C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage0\\bin\\cargo.exe" "rustc" "--manifest-path" "Cargo.toml" "-p" "path+file:///C:/a/_temp/msys64/tmp/.tmpflMHKG#match-stress@0.1.0" "--profile" "check" "--" "--wrap-rustc-with" "Eprintln"
[2024-10-31T10:07:06Z INFO  collector::compile::execute] run_rustc with incremental=true, profile=Check, scenario=Some(IncrFull), patch=None, backend=Llvm, phase=benchmark
[2024-10-31T10:07:06Z DEBUG collector::compile::execute] "\\\\?\\C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage0\\bin\\cargo.exe" "rustc" "--manifest-path" "Cargo.toml" "-p" "path+file:///C:/a/_temp/msys64/tmp/.tmpflMHKG#match-stress@0.1.0" "--profile" "check" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=C:\\a\\_temp\\msys64\\tmp\\.tmpflMHKG\\incremental-state"
[2024-10-31T10:07:08Z INFO  collector::compile::execute] run_rustc with incremental=true, profile=Check, scenario=Some(IncrUnchanged), patch=None, backend=Llvm, phase=benchmark
[2024-10-31T10:07:08Z DEBUG collector::compile::execute] "\\\\?\\C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage0\\bin\\cargo.exe" "rustc" "--manifest-path" "Cargo.toml" "-p" "path+file:///C:/a/_temp/msys64/tmp/.tmpflMHKG#match-stress@0.1.0" "--profile" "check" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=C:\\a\\_temp\\msys64\\tmp\\.tmpflMHKG\\incremental-state"
[2024-10-31T10:07:08Z DEBUG collector::compile::benchmark] Benchmark iteration 1/1
[2024-10-31T10:07:08Z INFO  collector::compile::execute] run_rustc with incremental=false, profile=Debug, scenario=Some(Full), patch=None, backend=Llvm, phase=benchmark
[2024-10-31T10:07:08Z DEBUG collector::compile::execute] "\\\\?\\C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage0\\bin\\cargo.exe" "rustc" "--manifest-path" "Cargo.toml" "-p" "path+file:///C:/a/_temp/msys64/tmp/.tmpP3W1rJ#match-stress@0.1.0" "--" "--wrap-rustc-with" "Eprintln"
[2024-10-31T10:07:10Z INFO  collector::compile::execute] run_rustc with incremental=true, profile=Debug, scenario=Some(IncrFull), patch=None, backend=Llvm, phase=benchmark
---
[2024-10-31T10:07:31Z DEBUG collector::compile::benchmark] Benchmark iteration 1/1
[2024-10-31T10:07:31Z INFO  collector::compile::execute] run_rustc with incremental=false, profile=Debug, scenario=Some(Full), patch=None, backend=Llvm, phase=benchmark
[2024-10-31T10:07:31Z DEBUG collector::compile::execute] "\\\\?\\C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage0\\bin\\cargo.exe" "rustc" "--manifest-path" "Cargo.toml" "-p" "path+file:///C:/a/_temp/msys64/tmp/.tmpRKcYCA#tuple-stress@0.1.0" "--" "--wrap-rustc-with" "Eprintln"
[2024-10-31T10:07:34Z INFO  collector::compile::execute] run_rustc with incremental=true, profile=Debug, scenario=Some(IncrFull), patch=None, backend=Llvm, phase=benchmark
[2024-10-31T10:07:34Z DEBUG collector::compile::execute] "\\\\?\\C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage0\\bin\\cargo.exe" "rustc" "--manifest-path" "Cargo.toml" "-p" "path+file:///C:/a/_temp/msys64/tmp/.tmpRKcYCA#tuple-stress@0.1.0" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=C:\\a\\_temp\\msys64\\tmp\\.tmpRKcYCA\\incremental-state"
[2024-10-31T10:07:37Z INFO  collector::compile::execute] run_rustc with incremental=true, profile=Debug, scenario=Some(IncrUnchanged), patch=None, backend=Llvm, phase=benchmark
[2024-10-31T10:07:37Z DEBUG collector::compile::execute] "\\\\?\\C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage0\\bin\\cargo.exe" "rustc" "--manifest-path" "Cargo.toml" "-p" "path+file:///C:/a/_temp/msys64/tmp/.tmpRKcYCA#tuple-stress@0.1.0" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=C:\\a\\_temp\\msys64\\tmp\\.tmpRKcYCA\\incremental-state"
[2024-10-31T10:07:38Z DEBUG collector::compile::benchmark::patch] applying new row to "C:\\a\\_temp\\msys64\\tmp\\.tmpRKcYCA"
[2024-10-31T10:07:38Z INFO  collector::compile::execute] run_rustc with incremental=true, profile=Debug, scenario=Some(IncrPatched), patch=Some(Patch { index: 0, name: PatchName("new row"), path: "0-new-row.patch" }), backend=Llvm, phase=benchmark
[2024-10-31T10:07:38Z INFO  collector::compile::execute] run_rustc with incremental=true, profile=Debug, scenario=Some(IncrPatched), patch=Some(Patch { index: 0, name: PatchName("new row"), path: "0-new-row.patch" }), backend=Llvm, phase=benchmark
[2024-10-31T10:07:38Z DEBUG collector::compile::execute] "\\\\?\\C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage0\\bin\\cargo.exe" "rustc" "--manifest-path" "Cargo.toml" "-p" "path+file:///C:/a/_temp/msys64/tmp/.tmpRKcYCA#tuple-stress@0.1.0" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=C:\\a\\_temp\\msys64\\tmp\\.tmpRKcYCA\\incremental-state"
[2024-10-31T10:07:42Z DEBUG collector::compile::benchmark] Benchmark iteration 1/1
[2024-10-31T10:07:42Z INFO  collector::compile::execute] run_rustc with incremental=false, profile=Opt, scenario=Some(Full), patch=None, backend=Llvm, phase=benchmark
[2024-10-31T10:07:42Z DEBUG collector::compile::execute] "\\\\?\\C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage0\\bin\\cargo.exe" "rustc" "--manifest-path" "Cargo.toml" "-p" "path+file:///C:/a/_temp/msys64/tmp/.tmpPRwVxz#tuple-stress@0.1.0" "--release" "--" "--wrap-rustc-with" "Eprintln"
[2024-10-31T10:07:45Z INFO  collector::compile::execute] run_rustc with incremental=true, profile=Opt, scenario=Some(IncrFull), patch=None, backend=Llvm, phase=benchmark
---
   Compiling rustc_driver v0.0.0 (C:\a\rust\rust\compiler\rustc_driver)
[RUSTC-TIMING] rustc_driver test:false 4.598
error: linking with `link.exe` failed: exit code: 1104
  |
  = note: "C:\\Program Files\\Microsoft Visual Studio\\2022\\Enterprise\\VC\\Tools\\MSVC\\14.41.34120\\bin\\HostX64\\x64\\link.exe" "/NOLOGO" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\symbols.o" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\deps\\rustc_main-fb6579e0c7dac032.rustc_main.cc37c2271c07455e-cgu.0.rcgu.o" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\deps\\rustc_driver-697bdd72392c31d0.dll.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libcompiler_builtins-312cc7332d5665e0.rlib" "psapi.lib" "shell32.lib" "ole32.lib" "uuid.lib" "advapi32.lib" "ws2_32.lib" "ntdll.lib" "kernel32.lib" "advapi32.lib" "ole32.lib" "oleaut32.lib" "advapi32.lib" "cfgmgr32.lib" "gdi32.lib" "kernel32.lib" "msimg32.lib" "opengl32.lib" "user32.lib" "winspool.lib" "bcrypt.lib" "advapi32.lib" "legacy_stdio_definitions.lib" "kernel32.lib" "kernel32.lib" "advapi32.lib" "ntdll.lib" "userenv.lib" "ws2_32.lib" "dbghelp.lib" "/defaultlib:libcmt" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\advapi32.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\api-ms-win-core-errorhandling-l1-1-3.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\api-ms-win-core-file-fromapp-l1-1-0.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\api-ms-win-core-handle-l1-1-0.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\api-ms-win-core-ioring-l1-1-0.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\api-ms-win-core-memory-l1-1-3.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\api-ms-win-core-memory-l1-1-4.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\api-ms-win-core-memory-l1-1-5.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\api-ms-win-core-memory-l1-1-6.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\api-ms-win-core-memory-l1-1-7.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\api-ms-win-core-memory-l1-1-8.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\api-ms-win-core-synch-l1-2-0.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\api-ms-win-core-sysinfo-l1-2-0.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\api-ms-win-core-sysinfo-l1-2-3.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\api-ms-win-core-sysinfo-l1-2-4.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\api-ms-win-core-sysinfo-l1-2-6.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\api-ms-win-core-util-l1-1-1.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\api-ms-win-core-winrt-error-l1-1-0.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\api-ms-win-core-winrt-l1-1-0.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\api-ms-win-core-wow64-l1-1-1.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\api-ms-win-security-base-l1-2-2.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\avrt.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\bcp47mrm.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\bcryptprimitives.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\clfsw32.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\dbghelp.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\elscore.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\gdi32.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\icu.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\imagehlp.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\kernel32.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\ktmw32.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\netapi32.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\normaliz.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\ntdll.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\ntdllk.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\ole32.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\oleacc.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\oleaut32.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\propsys.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\psapi.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\rtworkq.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\txfw32.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\user32.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\usp10.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\version.dll_imports_indirect.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcYFtjIe\\wofutil.dll_imports_indirect.lib" "/NXCOMPAT" "/LIBPATH:C:\\Program Files\\Microsoft Visual Studio\\2022\\Enterprise\\VC\\Tools\\MSVC\\14.41.34120\\atlmfc\\lib\\x64" "/LIBPATH:C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\stacker-1c496942f5526841\\out" "/LIBPATH:C:\\Program Files\\Microsoft Visual Studio\\2022\\Enterprise\\VC\\Tools\\MSVC\\14.41.34120\\atlmfc\\lib\\x64" "/LIBPATH:C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\psm-b307fe5d9726f8b1\\out" "/LIBPATH:C:\\Program Files\\Microsoft Visual Studio\\2022\\Enterprise\\VC\\Tools\\MSVC\\14.41.34120\\atlmfc\\lib\\x64" "/LIBPATH:C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\blake3-4a372efeb6f355a1\\out" "/LIBPATH:C:\\Program Files\\Microsoft Visual Studio\\2022\\Enterprise\\VC\\Tools\\MSVC\\14.41.34120\\atlmfc\\lib\\x64" "/LIBPATH:C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\blake3-4a372efeb6f355a1\\out" "/LIBPATH:C:\\Program Files\\Microsoft Visual Studio\\2022\\Enterprise\\VC\\Tools\\MSVC\\14.41.34120\\atlmfc\\lib\\x64" "/LIBPATH:C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc_llvm-e2958cd171cbaa3b\\out" "/LIBPATH:C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\llvm\\lib" "/OUT:C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\deps\\rustc_main-fb6579e0c7dac032.exe" "/OPT:REF,ICF" "/DEBUG" "/PDBALTPATH:%_PDB%" "/MANIFEST:EMBED" "/MANIFESTINPUT:C:\\a\\rust\\rust\\compiler\\rustc\\Windows Manifest.xml" "/WX"
  = note: LINK : fatal error LNK1104: cannot open file 'C:\a\rust\rust\build\x86_64-pc-windows-msvc\stage1-rustc\x86_64-pc-windows-msvc\release\deps\rustc_main-fb6579e0c7dac032.exe'␍

[RUSTC-TIMING] rustc_main test:false 0.643
error: could not compile `rustc-main` (bin "rustc-main") due to 1 previous error
Build completed unsuccessfully in 0:07:36
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/11609160216/job/32328589939)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/11609160216/job/32328589939","builder_name":"checks-actions"} -->

jieyouxu commented:
@bors retry (msvc spurious linker failure)

klensy commented:
Maybe don't rollup pr where optimization was disabled?

DianQK commented:
> Maybe don't rollup pr where optimization was disabled?

cc @matthiaskrgr

matthiaskrgr commented:
ah, didn't see that this affected default mir opt levels
@bors rollup=never
shoud have been set then instead of iffy :)

jieyouxu commented:
Oh oops should have rollup=never yeah.

apiraino commented:
Beta backport accepted as per compiler team [on Zulip](https://rust-lang.zulipchat.com/#narrow/channel/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202024-10-31/near/479887293), once it gets merged. A backport PR will be authored by the release team at the end of the current development cycle.

@rustbot label +beta-accepted

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/11614945338/job/32352334756)
Approved by: cjgillot,DianQK
Pushing a0d98ff0e5b6e1f2c63fd26f68484792621b235c to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"cjgillot,DianQK","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/11614945338/job/32352334756"},"merge_sha":"a0d98ff0e5b6e1f2c63fd26f68484792621b235c"} -->

rust-timer commented:
Finished benchmarking commit (a0d98ff0e5b6e1f2c63fd26f68484792621b235c): [comparison URL](https://perf.rust-lang.org/compare.html?start=20c909ff9cdd88d33768a4ddb8952927a675b0ad&end=a0d98ff0e5b6e1f2c63fd26f68484792621b235c&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - please read the text below


Our benchmarks found a performance regression caused by this PR.
This might be an actual regression, but it can also be just noise.

**Next Steps**:

- If the regression was expected or you think it can be justified,
please write a comment with sufficient written justification, and add
`@rustbot label: +perf-regression-triaged` to it, to mark the regression as triaged.
- If you think that you know of a way to resolve the regression, try to create
a new PR with a fix for the regression.
- If you do not understand the regression or you think that it is just noise,
you can ask the `@rust-lang/wg-compiler-performance` working group for help (members of this group
were already notified of this PR).

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance



### [Instruction count](https://perf.rust-lang.org/compare.html?start=20c909ff9cdd88d33768a4ddb8952927a675b0ad&end=a0d98ff0e5b6e1f2c63fd26f68484792621b235c&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.7%  | [0.3%, 1.9%]   | 14    |
| Regressions ❌ <br /> (secondary)  | 0.7%  | [0.2%, 2.3%]   | 16    |
| Improvements ✅ <br /> (primary)   | -0.3% | [-0.3%, -0.3%] | 1     |
| Improvements ✅ <br /> (secondary) | -0.3% | [-0.4%, -0.3%] | 5     |
| All ❌✅ (primary)                 | 0.7%  | [-0.3%, 1.9%]  | 15    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=20c909ff9cdd88d33768a4ddb8952927a675b0ad&end=a0d98ff0e5b6e1f2c63fd26f68484792621b235c&stat=max-rss)
<details>
<summary>Results (primary -2.2%, secondary 0.6%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 3.9%  | [3.5%, 4.3%]   | 2     |
| Regressions ❌ <br /> (secondary)  | 2.7%  | [1.9%, 4.5%]   | 4     |
| Improvements ✅ <br /> (primary)   | -6.2% | [-9.1%, -3.5%] | 3     |
| Improvements ✅ <br /> (secondary) | -2.2% | [-2.6%, -2.0%] | 3     |
| All ❌✅ (primary)                 | -2.2% | [-9.1%, 4.3%]  | 5     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=20c909ff9cdd88d33768a4ddb8952927a675b0ad&end=a0d98ff0e5b6e1f2c63fd26f68484792621b235c&stat=cycles:u)
<details>
<summary>Results (primary 0.9%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.9% | [0.9%, 0.9%] | 2     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.9% | [0.9%, 0.9%] | 2     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=20c909ff9cdd88d33768a4ddb8952927a675b0ad&end=a0d98ff0e5b6e1f2c63fd26f68484792621b235c&stat=size:linked_artifact)
<details>
<summary>Results (primary 0.1%, secondary 0.3%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.2%  | [0.0%, 1.3%]   | 25    |
| Regressions ❌ <br /> (secondary)  | 0.6%  | [0.2%, 1.2%]   | 12    |
| Improvements ✅ <br /> (primary)   | -0.1% | [-0.3%, -0.0%] | 11    |
| Improvements ✅ <br /> (secondary) | -0.0% | [-0.0%, -0.0%] | 10    |
| All ❌✅ (primary)                 | 0.1%  | [-0.3%, 1.3%]  | 36    |
</details>

**Bootstrap**: 784.059s -> 782.659s (-0.18%)
**Artifact size**: 333.59 MiB -> 333.43 MiB (-0.05%)

lqd commented:
This deactivated an unsound MIR optimization until it is fixed. 

@rustbot label: +perf-regression-triaged
